Add class to track metadata of local files

dafyddstephenson opened the pull request:
> WIP! Opening as a draft to solicit feedback on the high-level structure

EDIT: ready to go!
NOTE: this is built on top of the commits in #288 , merge #288 first and then update diffs.


- [ ] Closes #xxxx
- [x] Tests added
- [x] Tests passing
- [x] Full type hint coverage
- [x] Changes are documented in `docs/releases.rst`
- [x] New functions/methods are listed in `api.rst`
- [x] New functionality has documentation

As part of consolidating the various `get()` and `get()`-like methods on classes like `AdditionalCode`, `InputDataset`, etc. into a single `DataSource.get()` method, this PR first consolidates related attributes _set_ by the `get()`-like methods on these classes that are used to track local file metadata, e.g. `AdditionalCode._local_file_hash_cache` or `InputDataset._local_file_stat_cache`.

This PR introduces a class `LocalFileStatistics` to track modification times, filesizes, SHA256 hashes, and names for a selection of (or single) co-located files. For now an instance of this class will be assigned to an attribute on the various "getting" classes, and they will keep their existing `get()` methods. In the future, `DataSource.get()` will return a `LocalFileStatistics` instance which can then optionally be assigned to an attribute on the class in question.

Example :

```
In [1]: from cstar.base import AdditionalCode

In [2]: compile_time_code = AdditionalCode(location = "~/Code/my_ucla_roms/Examples/Rivers_real/",
                                   files=["cppdefs.opt",
                                          "flux_frc.opt",
                                          "ocean_vars.opt",
                                          "param.opt",
                                          "river_frc.opt"])

In [3]: print(compile_time_code.local_file_stats)
None

In [4]: compile_time_code.get("tmp/")
[INFO] • Copying cppdefs.opt to /Users/dafyddstephenson/Code/my_c_star/cstar/base/tmp
[INFO] • Copying flux_frc.opt to /Users/dafyddstephenson/Code/my_c_star/cstar/base/tmp
[INFO] • Copying ocean_vars.opt to /Users/dafyddstephenson/Code/my_c_star/cstar/base/tmp
[INFO] • Copying param.opt to /Users/dafyddstephenson/Code/my_c_star/cstar/base/tmp
[INFO] • Copying river_frc.opt to /Users/dafyddstephenson/Code/my_c_star/cstar/base/tmp
[INFO] ✅ All files copied successfully

In [5]: print(compile_time_code.local_file_stats)
LocalFileStatistics
-------------------
Parent directory: /Users/dafyddstephenson/Code/my_c_star/cstar/base/tmp
Files: 

Name                                     Hash         Size (bytes) Modified
cppdefs.opt                              65de992b17…  1202         2025-06-17 16:56:25
flux_frc.opt                             2600e40001…  1068         2025-06-17 16:56:25
ocean_vars.opt                           90fc1cf1e1…  2346         2025-06-17 16:56:25
param.opt                                48cc34acaa…  908          2025-06-17 16:56:25
river_frc.opt                            1f23e50bf4…  728          2025-06-17 16:56:25
```

Commit:
add LocalFileStatistics class and implement in AdditionalCode via local_file_stats attr

Commit:
reintroduce AdditionalCode.working_path as a property that returns LocalFileStatistics.parent_dir; fix pre-commit

Commit:
move all unit test mocks of Path.resolve() to a single fixture in conftest.py

Commit:
rebase onto refactored mock Path.resolve() and fix unit tests

Commit:
expand LocalFileStats usage into InputDataset and ROMSInputDataset

Commit:
fix symlink resolution bug causing integration tests to fail

Commit:
tidy up and fix existing unit test failures

Commit:
add unit tests for local_file_stats.py

Commit:
fill out docstrings

Commit:
docs

Commit:
missing typehints

ankona commented:
> > WIP! Opening as a draft to solicit feedback on the high-level structure
> > EDIT: ready to go!
> > NOTE: this is built on top of the commits in #288 , merge #288 first and then update diffs.

>Is this "draft only" comment still valid? Did you mean to be in draft mode and just clicked the wrong thing? 

No the only mistake here is the quote block should have ended before "EDIT". The PR was in draft mode until it wasn't

ankona commented on cstar/base/additional_code.py:
```diff
@@ -121,22 +123,20 @@ def __repr__(self) -> str:
         return self._checkout_target
 
+    @property
+    def working_path(self) -> Optional[Path]:
```
docstring

dafyddstephenson replied:
43f911d

ankona commented on cstar/base/additional_code.py:
```diff
@@ -121,22 +123,20 @@ def __repr__(self) -> str:
+        return self.local_file_stats.parent_dir
+
     @property
     def exists_locally(self):
```
dostring

dafyddstephenson replied:
43f911d

ankona commented on cstar/base/additional_code.py:
```diff
@@ -205,15 +205,15 @@ def get(self, local_dir: str | Path) -> None:
+            )
             self.log.info("✅ All files copied successfully")
-            self.working_path = local_dir
+            # self.working_path = local_dir
```
don't forget to remove your commented code

dafyddstephenson replied:
43f911d

ankona commented on cstar/base/additional_code.py:
```diff
@@ -77,8 +78,7 @@ def __init__(
         self._checkout_target = checkout_target
         self.files: Optional[list[str]] = [] if files is None else files
```
I think we might have a bug in here. Consider writing a test that puts in an empty string in the files list and we'll probably end up with a bad path.

You may end up doing something like:

```py
if any(f is None for f in files):
  raise ValueError("An invalid file path was supplied.")
```

ankona replied:
+1 for a test passing `files = [None, None]` producing an invalid state

dafyddstephenson replied:
017a2c8

ankona commented on cstar/base/additional_code.py:
```diff
@@ -121,22 +123,20 @@ def __repr__(self) -> str:
-        self.working_path (bool)"""
-        if (self.working_path is None) or (self._local_file_hash_cache is None):
+        if not self.local_file_stats:
+            return False
```
total style nitpick... consider a line of whitespace here to differentiate the if and try

dafyddstephenson replied:
43f911d

ankona commented on cstar/base/input_dataset.py:
```diff
@@ -85,94 +84,22 @@ def validate(self):
-
-        self._local_file_hash_cache.update(local_hash)
-        return local_hash
+    def working_path(self) -> Optional[Path | list[Path]]:
```
Consider simplifying consumer code by avoiding nullables and varying types.

```python
@property
def working_path(self) -> list[Path]:
    if not self.local_file_stats:
        return []

    return self.local_file_stats.paths
```

In addition... either the method is poorly named (working_**path** implies it's always a _singular_ result) or the return type allowing a list makes no sense.

If you want `working_path` to always return the first path, fine. Just give yourself another property `all_paths -> list[Path]` to expose the other things instead of having this property do double duty.

dafyddstephenson replied:
Opened #292 to address this. Bit of lore as to where this inconsistency came from in the first place in there.

ankona commented on cstar/base/input_dataset.py:
```diff
@@ -85,94 +84,22 @@ def validate(self):
-
-        self._local_file_hash_cache.update(local_hash)
-        return local_hash
+    def working_path(self) -> Optional[Path | list[Path]]:
```
docstring

dafyddstephenson replied:
a318d90

ankona commented on cstar/base/input_dataset.py:
```diff
@@ -74,9 +75,7 @@ def __init__(
-        self.working_path: Optional[Path | List[Path]] = None
-        self._local_file_hash_cache: Dict = {}
-        self._local_file_stat_cache: Dict = {}
+        self.local_file_stats: Optional[LocalFileStatistics] = None
```
Consider implementing `LocalFileStatistics` so you can avoid null value checks in client code. Using the same pattern as below, we can use it later without first checking for none...

Consider instantiating with:

```python
self.local_file_stats = LocalFileStatistics(
            paths=[],
            stats={},
            hashes={},
        )
```

You might need to tweak `LocalFileStatistics.is_valid` but you'll like using this better  if you can just do

```python
try:
  return self.local_file_stats.is_valid():
except ...:
   return False

...
```

dafyddstephenson replied:
I considered designing it as a `Collection` which would implicitly cover this, but in the end the PR was getting a bit big and taking a lot of time considering it's just a stepping stone to something else. I'll file a ticket once it's merged and revisit it once the subsequent PR closing #291 is in.

ankona commented on cstar/base/input_dataset.py:
```diff
@@ -283,6 +210,7 @@ def _symlink_or_download_from_source(
     ) -> str:
         if location_type == "path":
             source_location = Path(source_location).expanduser().resolve()
+            # TODO when refactoring get(), avoid calculating hash here
```
don't forget your todo. consider creating a ticket if it's not getting done in this PR

dafyddstephenson replied:
#291

ankona commented on cstar/base/local_file_stats.py:
```diff
@@ -0,0 +1,229 @@
+
+        Parameters
+        ----------
+        paths : list of Path
```
I just created a new ticket about these docstrings... The resulting documentation looks really bad and these parameters and properties are already covered in the documentation.

We should stop copy/pasting every docstring into the class level docstring. 

See [this](https://cworthy.atlassian.net/browse/CW-875) for an example of how poorly docs render when included in the class docstring.

dafyddstephenson replied:
API docs have been in need of a tidy-up for a while, so that would be great. I made a half-hearted effort once but didn't get far enough quick enough and bailed. If you could also add something to the contributing guide on how to write docstrings that get rendered nicely in the RTD that'd be great.

ankona commented on cstar/base/local_file_stats.py:
```diff
@@ -0,0 +1,229 @@
+
+        self.paths = [p.absolute() for p in paths]
+
+        if not all([d.parent == self.paths[0].parent for d in self.paths]):
```
```suggestion
        if not all(d.parent == self.paths[0].parent for d in self.paths):
```
unnecessary comprehension.

ankona commented on cstar/base/local_file_stats.py:
```diff
@@ -0,0 +1,229 @@
+    def __init__(
+        self,
+        paths: list[Path],
+        stats: Optional[dict[Path, os.stat_result]] = None,
```
Consider creating purpose-built methods for instantiating from pre-existing data. The constructor arguments are reduced and it makes it easier to understand.

```python
@classmethod
def hydrate(self, files: ..., stats: ..., hashes: ...) -> Self:
    if not files:
        raise ValueError("no files provided.")
    if not stats:
        raise ValueError("no stats provided.")
    if not hashes:
        raise ValueError("no hashes provided.")
    
    return LocalFileStats(files, stats, hashes)
```

Now, we have a method that doesn't allow nulls and it's really easy to separate when we should check for empty sets.

dafyddstephenson replied:
I think in this case it is simpler to maintain a single entry point with optionals for pre-computed values, rather than expanding the API with another constructor for a relatively uncommon case. Note also that in the current implementation, the `Optional`s aren't left as `None` in the final state.

ankona commented on cstar/base/local_file_stats.py:
```diff
@@ -0,0 +1,229 @@
+                )
+
+    def __str__(self) -> str:
+        base_str = self.__class__.__name__ + "\n"
```
these repeated string concatenations are poor form. Consider how much easier it is to see the intended formatting when avoiding doing your layout manually with `\n` and `+=`:

```python
return textwrap.dedent("""\
    {self.__class__.__name__}
    {"-" * (len(base_str) - 1)}
    Parent directory:
    {self.parent_dir}
    Files:
    
    {self.as_table()}""")
```

dafyddstephenson replied:
This choice was made to allow conditionals in the `__str__` construction. There aren't any here, but in many other cases, lines are added or not depending on whether certain attributes are set, making a cleaner final `__str__` representation.

ScottEilerman replied:
For what it's worth, something like `f"{'foobar\n' if True else ''}"` is totally valid, though if you use a full dedent block like above, I guess you'd still end up with an empty line unless you pack optional stuff on the line above it which would get messy.

I'm also wondering if a lot of these str/reprs will be largely unused or unseen once we move away from interactive notebooks and more toward automated processes, so maybe not worth sinking a lot of time into until we see whether we really want them later.

ankona commented on cstar/base/local_file_stats.py:
```diff
@@ -0,0 +1,229 @@
+        return base_str
+
+    def __repr__(self) -> str:
+        repr_str = self.__class__.__name__ + "("
```
ditto on avoiding repetitive concatenations

dafyddstephenson replied:
See above comment

ankona commented on cstar/base/local_file_stats.py:
```diff
@@ -0,0 +1,229 @@
+
+        self.parent_dir = paths[0].parent
+
+        if stats is not None:
```
Same deal with nulls. Look for a way to process additions immediately to avoid having a null stats and hashes member (same as comment about instantiating the LocalFileStats to an empty version).

If we're exposing a list, like "self.files = []` and allowing people to modify it at-will, we put ourselves in this position where we can't trust our member variables. If we hide that as `self._files = []` we can expose 

```python
def add_files(self, flist: list[Path]) -> None:
   if not validate(f):
      raise ValueError
   self._files.extend(f.absolute() for f in flist)
```

When I notice that I'm making myself do repeated checks for nulls, i always start to ask "how can I avoid doing this everywhere?"

ankona commented on cstar/base/local_file_stats.py:
```diff
@@ -0,0 +1,229 @@
+        repr_str += ")"
+        return repr_str
+
+    def as_table(self, max_rows: int = 20) -> str:
```
As table is a "presentation class" and should not be included on this class, outright. Consider creating another class just to contain this in a more "UI-focused" package:

```
class Tabularizer:
  """A contract to spit out data that can be rendered in a table"""

  """A contract for entities that can be rendered in a tabular format"""
  def as_table(self) -> list[list[str]]:
    # this method returns lists of lists on purpose so i can separate how
    # i render and what i render
    ...
  def headers(self) -> list[str]:
    ...

class LocalFileStatisticsBriefTabularizer:
   """Tabularizer allows me to implement a specific data extraction""""
    def __init__(self, lfs: LocalFileStatistics) -> None:
        self._lfs = lfs

    def as_table(self) -> list[list[str]]:
        return [["a", "b"] for x in self._lfs]
    def headers(self) -> list[str]:
      return ["A HEAD","B HEAD"]

class LocalFileStatisticsFullTabularizer:
   """This Tabularizer allows me to avoid logic for another desired rendering...""""
    def __init__(self, lfs: LocalFileStatistics) -> None:
        self._lfs = lfs

    def as_table(self) -> list[list[str]]:
        return [["a", "b", "c", "d"] for x in self._lfs]  
    def headers(self) -> list[str]:
      return ["A HEAD","B HEAD","C HEAD","D HEAD"]    

class TableRenderer:
    """THe magic! Now, I can render ANY Tabularizer with one method..."""
    @classmethod
    def _csv(self, tt: Tabularizer) -> str:
        data = tt.as_table(). # i don't care what's in there anymore...
        shead = ','.join(tt.headers)
        sdata = ','.join(x for x in data)
        return f"{shead}\n{sdata}"

    @classmethod
    def _json(self, tt: Tabularizer) -> str:
        data = tt.as_table(). # i don't care what's in there anymore...
        header = tt.headers()
        d = dict(zip(header, x) for x in data)
        return json.dumps(d)

    @classmethod
    def render(self, format: Literal['json','csv'], tt: Tabularizer) -> str:
        fmt_map = {"csv": self._csv, "json": self._json}
        return fmt_map[format](tt)


.... 

# somewhere else, when i want to use it...
print(TableRenderer.render(LocalFileStatisticsFullTabularizer(lfs)))
        
```

We now have a really nice separation of concerns for building tabular data (even in multiple layouts) and actually rendering it. It's more testable, and we can avoid repeating table rendering code everywhere (improving output quality because it's more consistent).

dafyddstephenson replied:
There are other instances like this in the code and a dedicated issue/PR pair to address it would be worthwhile, but I'm going to leave that as separate to this PR.

ankona commented on cstar/roms/input_dataset.py:
```diff
@@ -268,21 +268,15 @@ def get(
-            self.log.info(f"⏭️ {self.working_path} already exists, skipping.")
-            return
+        if self.exists_locally:
+            assert self.local_file_stats is not None
```
assertion here is not documented in exceptions and is probably a bad choice. (also, consider rebasing off latest develop and using either `CStarError` or a new type).

You should raise a more descriptive error here than an AssertionError to inform the client code what went wrong

dafyddstephenson replied:
68888fb

ankona commented on cstar/roms/input_dataset.py:
```diff
@@ -268,21 +268,15 @@ def get(
-        if (self.exists_locally) and (working_path_parent == local_dir):
-            self.log.info(f"⏭️ {self.working_path} already exists, skipping.")
-            return
+        if self.exists_locally:
```
Consider inverting to avoid having nested ifs..

```python
if not self.exists_locally:
    self.warning(...)
    raise ValueError

if self.local_stats.parent_dir == local_dir:
    self.log.info(...)
    return
   ...

...
```

dafyddstephenson replied:
I don't think a guard clause makes sense here as `not self.exists_locally` doesn't require any special handling (like `raise`) - it's the expected situation, the correct result is to proceed without doing anything. I previously had the conditional as a one-liner (no nesting):

```
        if (self.exists_locally) and (self.local_file_stats.parent_dir == local_dir):  # type: ignore[union-attr]
            self.log.info(f"⏭️ {self.working_path} already exists, skipping.")
            return
```
... but find the current form more human readable, despite the single-level nest.

ankona commented on cstar/roms/input_dataset.py:
```diff
@@ -321,10 +315,11 @@ def _get_from_partitioned_source(
             new_np_xi=source_np_xi, new_np_eta=source_np_eta, parted_files=parted_files
         )
-        self.working_path = parted_files
+        # self.working_path = parted_files
```
don't forget your commented code

dafyddstephenson replied:
97769b4

ankona commented on cstar/roms/input_dataset.py:
```diff
@@ -321,10 +315,11 @@ def _get_from_partitioned_source(
-        self._local_file_stat_cache.update(self.partitioning._local_file_stat_cache)
-        self._local_file_hash_cache.update(self.partitioning._local_file_hash_cache)
+        self.local_file_stats = LocalFileStatistics(
+            paths=[p.resolve() for p in parted_files]
```
you should not be required to know that you have to `p.resolve` before passing data to the LFS object. this implies that you understand the internal workings of that object and, in fact, must know that in every place it's created (see all the .resolve calls below). 

Make LFS call resolve on what it receives if it wants that?

dafyddstephenson replied:
[901abc0](https://github.com/CWorthy-ocean/C-Star/pull/285/commits/901abc02b0a80200838fa2db908540c383d6c9af) - This was made redundant a number of commits ago and I just forgot to tidy it up

ankona commented on cstar/roms/simulation.py:
```diff
@@ -1592,9 +1592,8 @@ def restart(self, new_end_date: str | datetime) -> "ROMSSimulation":
-            inp._local_file_hash_cache = None
-            inp._local_file_stat_cache = None
-            inp.working_path = None
+            inp.local_file_stats = None
```
Consider adding a `clear` method and maybe a `empty` property so we can avoid null checks.

`inp.clear()`

Same thought process here... you are making a `Simulation` responsible for what it means to be "empty" inside the input dataset, and that is not sustainable. Let the input_dataset own what clearing means.

dafyddstephenson replied:
c6c099f

Commit:
bugfix in mock and add assert in test_get_from_local_directory

Commit:
remove incorrect use of Path in MockInputDataset init

Commit:
assert mock_path_resolve called in test_get_when_file_exists

Commit:
add mock_path_resolve.assert_called() to tests that use this fixture in remaining InputDataset unit tests

Commit:
replace repeated Path definitions in test_local_hash_multiple_files with variables

Commit:
add mock_path_resolve.assert_called() checks throughout test_roms_input_dataset.py

Commit:
add remaining mock_path_resolve.assert_called() to test_input_dataset.py

Commit:
add remaining mock_path_resolve.assert_called() to test_input_dataset.py ii

Commit:
remove unused fixture in test_input_dataset.py

Commit:
Merge branch 'improved_mock_resolve' into local_file_stats

Commit:
docstrings and formatting changes in base/additional_code.py

Commit:
add check/unit test for falsy entries in AdditionalCode.files

Commit:
add missing docstrings to base/input_dataset.py

dafyddstephenson committed:
Remove unnecessary comprehension from base/local_file_stats.py

Co-authored-by: Chris McBride <3595025+ankona@users.noreply.github.com>

Commit:
replace assert with type: ignore in roms/input_dataset.py

Commit:
remove commented code

Commit:
add _clear method to InputDataset

Commit:
tidy up use of 'resolve' in roms/input_dataset.py re. local_file_stats

ScottEilerman commented on cstar/base/input_dataset.py:
```diff
@@ -269,9 +194,14 @@ def get(self, local_dir: str | Path) -> None:
+            ],
+            stats={p: p.stat()},
+            hashes={p: computed_file_hash},
+        )
```
I'm not sure if this is the same discussion as above that mentioned a Collection, but: seeing the usage here makes me wonder if there should be a `LocalFileStatistics` object for every path, and here we would hold a list of `LocalFileStatistics` objects. Then there's no need to have dictionaries for stats/hashes that you have to match with the list of paths, it's just 1:1:1 for every object.

I'll keep reading and see if I stick with this suggestion or if other usage will make me reconsider...

ScottEilerman replied:
After reading further, I do think this would be a good refactor, as it would simplify the `LocalFileStatistics` logic a lot by getting rid of some loops and potential null situations. Though you'd still need a collection construct to relocate those loops. Optional for now, it'll probably work fine as it is.

dafyddstephenson replied:
See below comment and 3a9846d6f578991a9c9df9ce321c446ce6687f3c

ScottEilerman commented on cstar/base/additional_code.py:
```diff
@@ -110,33 +118,41 @@ def __repr__(self) -> str:
+    def exists_locally(self) -> bool:
+        """Determines whether this AdditionalCode instance exists on the local
+        system."""
+        if not self.local_file_stats:
```
With regards to @ankona's comments about nulls throughout, I wonder how much use we have for the program state where we haven't done a `.get()` yet but don't want to just go ahead and do it. If we were to lazily `get()` anything whenever we end up needing it, it might reduce a lot of this "have I gotten it locally yet" logic because the answer would always be yes.

Food for thought for now. Maybe it will become apparent in the next `get()` PR if we should go down this road or not.

ScottEilerman reviewed (approved):
A few remarks but no blockers on my end.

Commit:
Refactor LocalFileStatistics:
- Add a FileInfo dataclass to hold stats and hashes without needing dicts
- Rename paths -> files in LocalFileStatistics which now accepts Sequence[Path | FileInfo] in __init__
- Add a LocalFileStatistics.__getitem__ method to lookup and return FileInfo for a given path
- Tidy up unit tests to pass, including simplifying some that failed due to overcomplicated Path-related mocks

dafyddstephenson commented:
@ScottEilerman I've refactored the class a little so it doesn't rely so heavily on cross-checking a bunch of dictionaries anymore, there's now a dataclass `FileInfo` that holds a single one:one:one path/stat/hash combination. I think having `LocalFileStatistics` manage multiple related files is best, but this lightweight dataclass kinda does the same thing as what you were suggesting with a one-instance-per-file arrangement. `LocalFileStatistics` now also has a `__getitem__` so you can do lookup via `my_local_file_stats[my_path]` and get back a `FileInfo` instance. I think it's a lot neater - thanks for the suggestion.

ScottEilerman committed:
example of more code on FileInfo class

ScottEilerman committed:
precommit, fix some types

Commit:
revert stats and hashes properties to return dict

Commit:
tidy up docstrings, replace tmp_file_pair fixture with factory to produce arbitrary number of tmp_files

Commit:
fill out unit tests

dafyddstephenson commented:
@ScottEilerman merged your branch and tidied up - thanks again!

Commit:
Merge branch 'main' into local_file_stats
